test/integration/http_integration.cc

```diff
@@ -1708,6 +1708,147 @@ void HttpIntegrationTest::testUpstreamProtocolError() {
   EXPECT_STREQ("503", response->headers().Status()->value().c_str());
 }
 
+void HttpIntegrationTest::testHeadersOnlyFilterEncoding() {
+  config_helper_.addFilter(R"EOF(
+name: encode-headers-only
+)EOF");
+  initialize();
+
+  codec_client_ = makeHttpConnection(lookupPort("http"));
+  auto response =
+      codec_client_->makeRequestWithBody(Http::TestHeaderMapImpl{{":method", "GET"},
+                                                                 {":path", "/test/long/url"},
+                                                                 {":scheme", "http"},
+                                                                 {":authority", "host"}},
+                                         128);
+  waitForNextUpstreamRequest();
+  upstream_request_->encodeHeaders(Http::TestHeaderMapImpl{{":status", "503"}}, false);
+  upstream_request_->encodeData(128, true);
```

Comment:
Can we make sure to test that if we don't encode a full body (but Envoy sends a full header-only response) that the upstream stream/connection gets reset?  ditto on the decode path

Reply:
Yeah sure. Do you happen to know how to explicitly test that? Is `waitForDisconnect` on the upstream req what you were thinking about?

Reply:
I think its waitForDisconnect/waitForReset based on upstream being H1/H2, like testRouterDownstreamDisconnectBeforeRequestComplete does.